### PR TITLE
feat(rome_js_analyze): `noUselessFragment`

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

It implements #3254 

The old rule had also a suggested fix, which I would like to implement in another PR. Having this PR with the correct diagnostics, would help me later to avoid regressions.

Also, I fixed some false positives in the `is_create_react_element` function. There was some logic missing.

It seems now that there are some recurrent patterns, I will refactor them in another PR.

## Test Plan

I added new test cases to cover the new rule and new test cases to cover the false positive I fixed in this PR.

### CODE REVIEW

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        let node = ctx.query();
+        let model = ctx.model();
+        match node {
+            NoUselessFragmentQuery::JsxFragment(fragment) => {
+                let parent = node.syntax().parent().and_then(|parent| {
+                    if let Some(parent) = JsxTagExpression::cast_ref(&parent) {
+                        parent.syntax().parent()
+                    } else {
+                        Some(parent)
+                    }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs:62]: Nit
```suggestion
                    match JsxTagExpression::try_cast(parent) {
                        Ok(parent) => parent.syntax().parent(),
                        Err(parent) => parent
                    } 
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                    parent_kind,
+                    Some(
+                        JsSyntaxKind::JS_RETURN_STATEMENT
+                            | JsSyntaxKind::JS_INITIALIZER_CLAUSE
+                            | JsSyntaxKind::JS_CONDITIONAL_EXPRESSION
+                            | JsSyntaxKind::JS_ARROW_FUNCTION_EXPRESSION
+                            | JsSyntaxKind::JS_FUNCTION_EXPRESSION
+                            | JsSyntaxKind::JS_FUNCTION_DECLARATION
+                    )
+                );
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs:76]: We should be able to move this block into the block where you extract the parent because it is never possible that the parent is any of this list if the "direct parent" isn't `JsxTagExpression`. 

```rust
let matches_allowed_parents = if let Some(expression) = node.parent::<JsxTagExpression>() {
  matches!(expression.syntax().parent().map(|p| p.kind()), .... )
} else {
  false 
}
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                let child_list = fragment.children();
+                if child_list.len() <= 1 && !matches_allowed_parents {
+                    return if child_list.is_empty() {
+                        Some(NoUselessFragmentState::Empty)
+                    } else {
+                        // SAFETY: it has at least one item
+                        let attribute = child_list.first().unwrap();
+                        Some(NoUselessFragmentState::Attribute(attribute))
+                    };
+                }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs:87]: Nit: You an avoid the `unwrap` by doing

```rust
match child_list.first() {
  Some(first) if child_list.len() == 1 => {
  Some(NoUselessFragmentState::Attribute(first))
  },
  None => Some(NoUselessFragmentState::Empty),
  _ => None
}
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                    )
+                );
+
+                let child_list = fragment.children();
+                if child_list.len() <= 1 && !matches_allowed_parents {
+                    return if child_list.is_empty() {
+                        Some(NoUselessFragmentState::Empty)
+                    } else {
+                        // SAFETY: it has at least one item
+                        let attribute = child_list.first().unwrap();
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs:84]: What's the meaning of `attribute` in this context? If I think of attributes in the jsx context than I think of `<div attribute>`

Comment (ematipico) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_useless_fragment.rs:84]: Yeah that was some old naming, sorry. It should be "child", not attribute

File: crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    <>foo</>
+    <React.Fragment>foo</React.Fragment>
+    <Fragment>foo</Fragment>
+</>
+
+```
+
+# Diagnostics
+```
+warning[nursery/noUselessFragment]: Avoid using unnecessary Fragment.
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap:17]: should *Fragment* be plural? Avoid using unnecessary Fragments?

File: crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  ┌─ withChildren.jsx:2:5
+  │
+2 │     <>foo</>
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap:20]: Is `Fragment` jargon? Is it OK for us to assume that people know that `<>` is called a fragment or should we change the messaging

File: crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap [hunk 7]
```diff
@@ -18,7 +18,7 @@ createElement('div', {
 # Diagnostics
 ```
 warning[nursery/noDangerouslySetInnerHtml]: Avoid passing content using the dangerouslySetInnerHTML prop.
-  ┌─ createElementBinding.js:4:5
+  ┌─ createElementBindingInvalid.js:4:5
```
Comment (MichaReiser) [inline crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap:21]: Would you mind adding an example with a comment (and yes, this is valid)

```javascript
</* test */>empty</>;
<>empty</* comment */ />;
<>empty</ /* comment */>
```

I would argue that the rule should probably not trigger for these cases.

**Review by MichaReiser — APPROVED** [review]

File: crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap (hunk 5 above)
Comment (ematipico) [reply crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap:17]: Yeah it makes sense. I also found out that we don't have a guideline of when using singular and plurals inside a rule.

File: crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap (hunk 7 above)
Comment (ematipico) [reply crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap:21]: The rule as it is now triggers for these cases, it seems. Would you mind give us a reason, please? It seems an edge case, so I can better document the code and rule itself

File: crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap (hunk 6 above)
Comment (ematipico) [reply crates/rome_js_analyze/tests/specs/nursery/noUselessFragment/withChildren.jsx.snap:20]: We will have a rule called [`useFragmentSyntax`](https://github.com/rome/tools/blob/archived-js/website/src/docs/lint/rules/react/useFragmentSyntax.md). Maybe that's where we can give more explanation?

File: crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap (hunk 7 above)
Comment (MichaReiser) [reply crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap:21]: My reasoning is that these fragments are entirely useless because it is otherwise not possible to add the comments (except using an expression). 

However, we could reason that we should flag these nonetheless but not provide an autofix because it is unclear where to place the comment?

Comment (ematipico) [reply crates/rome_js_analyze/tests/specs/nursery/noDangerouslySetInnerHtml/createElementBindingInvalid.js.snap:21]: Yeah this rule is not over yet, there are still few cases that doesn't take into account. And the autofix will arrive with another PR :)

**Comment by ematipico** [comment]
@leops the last commit adds a fix and test case to cover what you suggested. Nice catch!

**Review by leops — APPROVED** [review]
